### ISSUE TITLE
HA Control plane does not work locally on Ubuntu with kind, CAPD and kind images >= 1.21.1

### ISSUE THREAD

**Issue by johannesfrey** [kubernetes-sigs/cluster-api#6120]
**What steps did you take and what happened:**
* Following the [Rapid iterative development with Tilt](https://cluster-api.sigs.k8s.io/developer/tilt.html) section.
* Deploying the [simple-cluster.yaml](https://github.com/kubernetes-sigs/cluster-api/blob/main/test/infrastructure/docker/examples/simple-cluster.yaml) resources but with a replica count of 3 for the [KubeadmControlPlane](https://github.com/kubernetes-sigs/cluster-api/blob/main/test/infrastructure/docker/examples/simple-cluster.yaml#L46) resource.

After following the above steps only a single control plane node/container comes up. The second replica fails with the systemd kubelet service crash looping with the following error message:
```
"Failed to run kubelet" err="failed to run Kubelet: invalid configuration: cgroup-root [\"kubelet\"] doesn't exist"
```
There have been several discussions and a PR regarding this issue, with a temporary fix to pin kubelet's cgroup driver to `cgroupfs` :
* https://github.com/kubernetes-sigs/cluster-api/pull/4614
* https://kubernetes.slack.com/archives/C8TSNPY4T/p1619462894183700
* https://kubernetes.slack.com/archives/C8TSNPY4T/p1638448976306000

As I mentioned above. A single replica works fine. Also if the kind image is changed to something below 1.21.1 (e.g. 1.20.7) all 3 control plane replicas come up just fine.

**What did you expect to happen:**
* That all 3 control plane nodes come up

**Anything else you would like to add:**
As comparison, I also created a a HA control plane cluster manually with `kind create cluster`. This way all control plane nodes come up without the above mentioned kubelet error:
* Command: `kind create cluster --name test --config /tmp/kind-cluster-cgroupfs.yaml --image='kindest/node:v1.23.3'`
* Config:
  ```
  kind: Cluster
  apiVersion: kind.x-k8s.io/v1alpha4
  kubeadmConfigPatches:
  - |
    apiVersion: kubelet.config.k8s.io/v1beta1
    kind: KubeletConfiguration
    evictionHard:
      nodefs.available: "0%"
    cgroupDriver: "cgroupfs"
  nodes:
  - role: control-plane
  - role: control-plane
  - role: control-plane
  - role: worker
  - role: worker
  - role: worker
  ```

**Environment:**

- Cluster-api version: Current [main](https://github.com/kubernetes-sigs/cluster-api/commit/cca725c1ca50ee58ca1d4531dbeb2bb7a92669c7)
- Minikube/KIND version: kind v0.11.1
- Kubernetes version: (use `kubectl version`):
  ```
  Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.7", GitCommit:"1f86634ff08f37e54e8bfcd86bc90b61c98f84d4", GitTreeState:"clean", BuildDate:"2021-11-17T14:41:19Z", GoVersion:"go1.16.10", Compiler:"gc", Platform:"linux/amd64"}
  Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-21T23:01:33Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}
  ```
- OS (e.g. from `/etc/os-release`):
  ```
  NAME="Ubuntu"
  VERSION="20.04.3 LTS (Focal Fossa)"
  ID=ubuntu
  ID_LIKE=debian
  PRETTY_NAME="Ubuntu 20.04.3 LTS"
  VERSION_ID="20.04"
  HOME_URL="https://www.ubuntu.com/"
  SUPPORT_URL="https://help.ubuntu.com/"
  BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
  PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
  VERSION_CODENAME=focal
  UBUNTU_CODENAME=focal
  ```

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]
 
<sub>Johannes Frey <[johannes.frey@daimler.com](mailto:johannes.frey@daimler.com)>, Daimler TSS GmbH ([Provider Information](https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md))</sub>

**Comment by killianmuldoon** [comment]
@johannesfrey Have you got the CAPD logs for when this issue happens?

**Comment by johannesfrey** [comment]
Sure, but it is pretty verbose. I stripped out `bootstrap data`.

TLDR: `[kubelet-check] It seems like the kubelet isn't running or healthy.\n[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get \"http://localhost:10248/healthz\": dial tcp [::1]:10248: connect: connection refused.`

Full logs:
```
I0214 10:51:51.962221       9 machine.go:362] controller/dockermachine "msg"="Failed running command" "name"="controlplane-58h9q" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="DockerMachine" "bootstrap data"="..." "command"={"Cmd":"/bin/sh","Args":["-c","kubeadm join --ignore-preflight-errors=all --config /run/kubeadm/kubeadm-join-config.yaml  && echo success > /run/cluster-api/bootstrap-success.complete"],"Stdin":""} "stderr"="\t[WARNING SystemVerification]: failed to parse kernel config: unable to load kernel module: \"configs\", output: \"modprobe: FATAL: Module configs not found in directory /lib/modules/5.4.0-99-generic\\n\", err: exit status 1\nerror execution phase kubelet-start: error uploading crisocket: nodes \"my-cluster-controlplane-7c5r2\" not found\nTo see the stack trace of this error execute with --v=5 or higher\n" "stdout"="[preflight] Running pre-flight checks\n[preflight] The system verification failed. Printing the output from the verification:\n\u001b[0;37mKERNEL_VERSION\u001b[0m: \u001b[0;32m5.4.0-99-generic\u001b[0m\n\u001b[0;37mOS\u001b[0m: \u001b[0;32mLinux\u001b[0m\n\u001b[0;37mCGROUPS_CPU\u001b[0m: \u001b[0;32menabled\u001b[0m\n\u001b[0;37mCGROUPS_CPUACCT\u001b[0m: \u001b[0;32menabled\u001b[0m\n\u001b[0;37mCGROUPS_CPUSET\u001b[0m: \u001b[0;32menabled\u001b[0m\n\u001b[0;37mCGROUPS_DEVICES\u001b[0m: \u001b[0;32menabled\u001b[0m\n\u001b[0;37mCGROUPS_FREEZER\u001b[0m: \u001b[0;32menabled\u001b[0m\n\u001b[0;37mCGROUPS_MEMORY\u001b[0m: \u001b[0;32menabled\u001b[0m\n\u001b[0;37mCGROUPS_PIDS\u001b[0m: \u001b[0;32menabled\u001b[0m\n\u001b[0;37mCGROUPS_HUGETLB\u001b[0m: \u001b[0;32menabled\u001b[0m\n[preflight] Reading configuration from the cluster...\n[preflight] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -o yaml'\n[preflight] Running pre-flight checks before initializing the new control plane instance\n[preflight] Pulling images required for setting up a Kubernetes cluster\n[preflight] This might take a minute or two, depending on the speed of your internet connection\n[preflight] You can also perform this action in beforehand using 'kubeadm config images pull'\n[certs] Using certificateDir folder \"/etc/kubernetes/pki\"\n[certs] Generating \"apiserver\" certificate and key\n[certs] apiserver serving cert is signed for DNS names [kubernetes kubernetes.default kubernetes.default.svc kubernetes.default.svc.cluster.local localhost my-cluster-controlplane-7c5r2] and IPs [10.96.0.1 172.18.0.10 172.18.0.4 127.0.0.1 0.0.0.0]\n[certs] Generating \"apiserver-kubelet-client\" certificate and key\n[certs] Generating \"front-proxy-client\" certificate and key\n[certs] Generating \"etcd/peer\" certificate and key\n[certs] etcd/peer serving cert is signed for DNS names [localhost my-cluster-controlplane-7c5r2] and IPs [172.18.0.10 127.0.0.1 ::1]\n[certs] Generating \"etcd/server\" certificate and key\n[certs] etcd/server serving cert is signed for DNS names [localhost my-cluster-controlplane-7c5r2] and IPs [172.18.0.10 127.0.0.1 ::1]\n[certs] Generating \"etcd/healthcheck-client\" certificate and key\n[certs] Generating \"apiserver-etcd-client\" certificate and key\n[certs] Valid certificates and keys now exist in \"/etc/kubernetes/pki\"\n[certs] Using the existing \"sa\" key\n[kubeconfig] Generating kubeconfig files\n[kubeconfig] Using kubeconfig folder \"/etc/kubernetes\"\n[kubeconfig] Writing \"admin.conf\" kubeconfig file\n[kubeconfig] Writing \"controller-manager.conf\" kubeconfig file\n[kubeconfig] Writing \"scheduler.conf\" kubeconfig file\n[control-plane] Using manifest folder \"/etc/kubernetes/manifests\"\n[control-plane] Creating static Pod manifest for \"kube-apiserver\"\n[control-plane] Creating static Pod manifest for \"kube-controller-manager\"\n[control-plane] Creating static Pod manifest for \"kube-scheduler\"\n[check-etcd] Checking that the etcd cluster is healthy\n[kubelet-start] Writing kubelet configuration to file \"/var/lib/kubelet/config.yaml\"\n[kubelet-start] Writing kubelet environment file with flags to file \"/var/lib/kubelet/kubeadm-flags.env\"\n[kubelet-start] Starting the kubelet\n[kubelet-start] Waiting for the kubelet to perform the TLS Bootstrap...\n[kubelet-check] Initial timeout of 40s passed.\n[kubelet-check] It seems like the kubelet isn't running or healthy.\n[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get \"http://localhost:10248/healthz\": dial tcp [::1]:10248: connect: connection refused.\n[kubelet-check] It seems like the kubelet isn't running or healthy.\n[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get \"http://localhost:10248/healthz\": dial tcp [::1]:10248: connect: connection refused.\n[kubelet-check] It seems like the kubelet isn't running or healthy.\n[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get \"http://localhost:10248/healthz\": dial tcp [::1]:10248: connect: connection refused.\n[kubelet-check] It seems like the kubelet isn't running or healthy.\n[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get \"http://localhost:10248/healthz\": dial tcp [::1]:10248: connect: connection refused.\n[kubelet-check] It seems like the kubelet isn't running or healthy.\n[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get \"http://localhost:10248/healthz\": dial tcp [::1]:10248: connect: connection refused.\n"
```

**Comment by killianmuldoon** [comment]
Is there anything non kubelet specific in the Docker logs? I'm just wondering if something with the docker container has gone wrong that looks like an unhealthy kubelet to help triage where the issue is.

**Comment by johannesfrey** [comment]
> Is there anything non kubelet specific in the Docker logs? I'm just wondering if something with the docker container has gone wrong that looks like an unhealthy kubelet to help triage where the issue is.

Unfortunately, I'm not seeing anything other suspicious. But I think I found the "issue":

If I look at the file `/var/lib/kubelet/kubeadm-flags.env` of the first control plane node (the one that does work). It looks like this:
```
KUBELET_KUBEADM_ARGS="--cgroup-driver=cgroupfs --container-runtime=remote --container-runtime-endpoint=/run/containerd/containerd.sock --eviction-hard=nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0% --pod-infra-container-image=k8s.gcr.io/pause:3.6"
```

The content of the second node (the one that does not come up with the above error) it looks like this:
```
KUBELET_KUBEADM_ARGS="--container-runtime=remote --container-runtime-endpoint=/run/containerd/containerd.sock --pod-infra-container-image=k8s.gcr.io/pause:3.6"
```

IIUIC this means that the second node will default to cgroupdriver systemd, which does not work with the hardcoded cgroup-root `/kubelet`. So it looks like that no kubeletExtraArgs are passed when other control-plane nodes do join the cluster, which makes sense, as those are only set in the [initConfiguration](https://github.com/kubernetes-sigs/cluster-api/blob/main/test/infrastructure/docker/examples/simple-cluster.yaml#L64). If I copy&paste `nodeRegistration` to an additional `joinConfiguration` everything works fine.

So it's not a bug but rather a missing configuration that I forgot to specify :sweat_smile:. I'm not sure, but does it make sense to also put a `joinConfiguration` in the example, even though it doesn't really make sense for a single replica. But at least it would prevent other users from running into this if they think they can just increment the replica count and be done with it :). WDYT?

**Comment by neolit123** [comment]
a maintainer recommends to use the latest kind image:
https://github.com/kubernetes-sigs/kind/issues/2625#issuecomment-1034755383

> cgroupDriver: "cgroupfs"

AFAIK if the host has cgroups v2 this driver will not work.

i have no explanation why single CP node works, but not multiple ones.

**Comment by neolit123** [comment]
> If I copy&paste nodeRegistration to an additional joinConfiguration everything works fine.

indeed, that is how kind patches work.

**Comment by johannesfrey** [comment]
> @johannesfrey Good finding. Let's add the joinconfiguration like we did in our cluster template (which is used by clusterctl):
> 
> https://github.com/kubernetes-sigs/cluster-api/blob/96936133d1bccd48eda535d74e82eea9babdea30/test/infrastructure/docker/templates/cluster-template-development.yaml#L49-L64
> 
> Let's also fix the other example YAMLs.
> 
> Can you / do you want to open a PR for it?

Sure, I will open a PR.
